feat(@clayui/layout): create low-level layout utilities

bryceosterhaus opened the pull request:
Here is my initial idea for the low-level components of row, col, and container.

cc @pat270 can you checkout the demos on storybook and verify these make sense?

jbalsas commented on packages/clay-layout/src/Col.tsx:
```diff
@@ -0,0 +1,85 @@
+	sm,
+	xs,
+	...otherProps
+}) => {
```
Nice! 

We didn't add `align` and `order` to our JSP implementation... might be worth doing based on this! 🤔

bryceosterhaus replied:
My goal was to mimic the docs examples as much as possible without having the user to have to add custom classes.

marcoscv-work replied:
We even don't recommend to use `order` since it causes accessibility problems of tab perception, I would not add it to avoid encouraging its use

jbalsas commented on packages/clay-layout/src/Container.tsx:
```diff
@@ -0,0 +1,61 @@
+	/**
+	 * Adds `.container-fluid-${size}` class to set max width on container.
+	 */
+	fluidSize?: 'sm' | 'md' | 'lg' | 'xl';
```
We just [called this size](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ContainerTag.java#L160) in our JSP tag since it seems to be the basic function of the Container, so why the prefix?

We also default to `xl` since it's by far the main usage we saw across the board... might want to consider it here too

bryceosterhaus replied:
I used the prefix since I didn't want people to think `size` could correlate to `container-${size}`. I thin by making it explicitly named `fluid` its apparent that its for the `container-fluid` class

marcoscv-work replied:
I think the correct name properties should be `fluidMaxSize` since we only use classes `container-fluid-max-*breakpoint*` to create all Lexicon grids they need. On the other hand, I would recommend forgetting all rest of ways to create container size limit because there are a lot which probably a final customer could use but we do not need

jbalsas replied:
So... now that we have `container-fluid`... maybe this could be:

`fluid: boolean | ['sm', 'md', 'lg', 'xl']`?

jbalsas commented on packages/clay-layout/src/Container.tsx:
```diff
@@ -0,0 +1,61 @@
+	 * Adds `.container-fluid` class to create a fluid container that
+	 * doesn't expand beyond a set width
+	 */
+	fluid?: boolean;
```
I lost track of where, but we went through a back and forth over `To Fluid or Not To Fluid` and concluded that we didn't want To Fluid 😂 

So, by default, we'd have `<Container>...</Container>` output:

```html
<div class="container-fluid container-fluid-xl">
    ...
</div>
```

This is somewhat opinionated... but so is our design language and as long as this can be overridden we should be fine. Not sure we have a valid case for just `container`, at the very least is not widespread enough to grant a first-level API

bryceosterhaus replied:
Hmm, I don't think I was privy to that convo.

IMO, I think its really odd to default to `container-fluid` for a component called `Container`, especially when the class `container` is also a thing. 

I think it'd be fine to keep `container-fluid` as a default for the a taglib in portal, but for the Clay library I don't think it would make sense to default to fluid for the name `Container`

marcoscv-work replied:
Initially, my concern about the using of `container` when we already have a class called `container` was the same @bryceosterhaus already has wrote, after some days taking a look to what lexicon want and what we finally use, the uses cases are only a few, majority of them are the classes `container-fluid container-fluid-max-xl` and we don't use the `standard` bootstrap container.
So for us and right now the default container is the class combination: `container-fluid container-fluid-max-xl`. It still sounds bad to me but I have to admit our container is a no very semantic combination of classes,  maybe it's something we can study and try to improve in a future more semantic CSS.

IMHO, the `<Container>` output should be our standard and more used container

jbalsas commented on packages/clay-layout/src/Container.tsx:
```diff
@@ -0,0 +1,61 @@
+	 * between application controls and the form. This class only modifies
+	 * the padding on the container.
+	 */
+	formSize?: 'sm' | 'md' | 'lg' | 'xl';
```
This is actually an interesting take... we treated this as `className` but it might make sense as an API...

@eduardoallegrini, thoughts?

bryceosterhaus replied:
I went through our documentation and tried to expose API that we explicitly talk about in the css or in our layouts definition. This is why I added it.

marcoscv-work replied:
I think it can be confusing with form sizes, if we speak about 'padding or margin' this probably should be 'spacers'. We have 8 unit sizes defined:
https://github.com/liferay/clay/blob/0273c22d1f6a9dfbd22e39f9490d7a9da05e53b7/packages/clay-css/src/scss/atlas/variables/_globals.scss#L182-L211
It generates all famous helpers like: `m-1`, `pt-5`, and our won utilities helpers: `c-m-5`, `c-pt-5`, etc.

Info about forms size:

For forms we have two groups sizes, and it can be consider a container: the default one `form-group` and `form-group form-group-sm`

for individual forms (form-controls) we have three sizes:
![image](https://user-images.githubusercontent.com/7963804/81699352-814beb00-9467-11ea-886b-cdcc808cea50.png)

jbalsas replied:
I would remove this one, just on the basis that it's easier to add them later than to remove them.

We can implement for now using `className`, and can bring it as an API once the pattern is more clear?

jbalsas commented on packages/clay-layout/src/Container.tsx:
```diff
@@ -0,0 +1,61 @@
+	 * controls and view pages (e.g., Card View, Table View, or List View).
+	 * This class only modifies the padding on the container.
+	 */
+	view?: boolean;
```
Same as with `.formSize`...

@eduardoallegrini, thoughts?

kresimir-coko replied:
I'm unsure as to the importance of these "proprietary" classes if we're going with the approach of separating the library from Portal. If I were making an app using Clay, I wouldn't really bother too much with props that have these very specific use cases.

Perhaps the question to ask is: _"How often will the user use this component with pre-existing knowledge of which props they'll need?"_
If it were me, I'd use `<ClayContainer flexOrEquivalent />` and then build out the rest, and then when it came to styling I'd probably not go back and check every component to see if I can configure the layout using props but instead just work in CSS.

jbalsas replied:
> I were making an app using Clay, I wouldn't really bother too much with props that have these very specific use cases [...]
> 
> [...] I'd probably not go back and check every component to see if I can configure the layout using props but instead just work in CSS [...]

That begs the question... why would you bring in a library dependency such as Clay to substitute `<div>` by `<ClayContainer>` and keep it all the same?

If you won't bother checking the props and defaults of the library of your choosing, why'd you choose it in the first place? 

We've discussed at length our vetting process for picking dependencies that are actually worth their value. If we were to have that conversation around this library, maybe @wincent, or yourself would argue that we don't need Clay since we already have `flexOrEquivalent`.

That might mean, you wouldn't be a user of Clay in a normal situation, so we shouldn't model it around this preference :)

kresimir-coko replied:
The fact that the primary goal of Clay is to serve DXP makes my concern kinda obsolete. And that's fine, I was just looking out for the little man here 😄, all of us using Clay while building DXP will look into all the props and use them accordingly. So having the classes that are omnipresent throughout DXP  in this component as props is a time saver for us.

jbalsas replied:
> primary goal of Clay is to serve DXP

I didn't say this!!

**Primary goal of Clay is to make building Lexicon UIs easy, consistent, robust and safe.**

DXP is the main consumer right now, so we shouldn't fall into the common trap of serving imaginary users and failing to deliver to those teams. We've been down that path before. It ain't pretty 😉 

That being said, we also aim for other departments and teams to jump on board as Design practices converge. AC (Faro), IS, DXP Cloud... they will all benefit in the long run if we do a good job here. That's why looking out for those "general" users is still valid and important, just not at the expense of DXP.

kresimir-coko replied:
Yea, I was worried my wording was too strong. I get what you're saying 👍 . I was focusing on the imaginary users more than on DXP. That's what makes Lexicon great, as long as we make components based on their design it should be great for both DXP and the other projects Liferay has, right?

jbalsas replied:
> That's what makes Lexicon great, as long as we make components based on their design it should be great for both DXP and the other projects Liferay has, right?

That's the dream 😂

jbalsas replied:
Similar to `formSize`... I know this is being used, si this has indeed real backing, but I wouldn't commit to it too early.

jbalsas commented on packages/clay-layout/src/Row.tsx:
```diff
@@ -0,0 +1,49 @@
+	gutters = true,
+	justify,
+	...otherProps
+}) => {
```
We didn't add any of these to `<clay:col>` since we couldn't find real usages, so it might be premature API, but LGTM

bryceosterhaus replied:
I think its hard for devs to remember the util classes for alignment, my thought was by adding the align and justify props, it would be more accessible to devs via ts autocomplete.

jbalsas commented on packages/clay-layout/src/__tests__/__snapshots__/index.tsx.snap:
```diff
@@ -0,0 +1,146 @@
+exports[`ClayLayout renders 1`] = `
+<div>
+  <div
+    class="container"
```
As I said above, if we want to somewhat match expectations, we might want to be a tad bit opinionated here and have this be `container-fluid container-fluid-xl` which is by far our most beloved container 😂

bryceosterhaus replied:
I think I would push back on that for the library level. Unless you can think of a good naming convention for the component to indicate it will include those. For a user to use `<Container>` without props and then get those classes seems odd to me. On the portal side of things, we can default to that since it makes sense for that to be portal's default, but not sure it makes sense to default at the library level.

I would also potentially push back and add that if we expect the default behavior of container to be `container-fluid container-fluid-xl` we might consider making that at the css level.

kresimir-coko replied:
I agree with @bryceosterhaus here. Default values can be provided in the implementation on Portal, thus preserving abstraction for the end-user while still providing users of Clay as a library the flexibility.

jbalsas replied:
Hey @bryceosterhaus, @kresimir-coko, I'll try to go at some more length in the general comments, but:

Once again, I think we're failing to see that right now, 100% (unless Faro moved to Clay) of our users are DXP. Also, **90% of our usage when implementing lexicon is `container-fluid`.** This means that by not providing this we're refusing to provide **real value to real users** of our library based on some not-so-real user/library expectation. Since 90% of the times someone in Liferay looking to build a UI had to use `container-fluid`, it is our goal to make this easier for them.

Please, go through:
- https://github.com/jbalsas/liferay-portal/pull/2160/commits/e72f30c231e196b72dcf15bbff7d94ddda6b2655
- https://github.com/eduardoallegrini/liferay-portal/pull/71/commits
- https://github.com/eduardoallegrini/liferay-portal/pull/72/commits/4c9f07b95116fd6f27dba4f59e6e0557299bd29d

The main goal of Clay is to **simplify implementing Lexicon UIs**. The main implementor of Lexicon UIs is DXP. Based on the above changes, I'd say those are better defaults for implementing Lexicon UIs. Thus, those defaults belong in Clay, not in DXP. Otherwise, we're not making it any easier. There are literally thousands of occurrences of `container-fluid` and around tens of `container`. It doesn't make any sense that we'd favour the latter here.

If you want to be real purist here, **let's maybe create `FluidContainer` that composes `Container` and provides the right semantics and ergonomics to our devs**. We can do the same in the Portal side and mostly simply use `<clay:fluid-container>` and `<ClayLayout.FluidContainer>`.

jbalsas replied:
@wincent, @marcoscv-work, thoughts?

kresimir-coko replied:
@jbalsas When you put it like that yea, I agree. I guess I didn't understand the exact goal of Clay, I thought it was a low-level library first, DXP-specific second. 

I figured people using DXP used Clay to build their apps, but I assume it's not like that after all. I don't personally care about being a purist here, so I have no problem with having the default be `container-fluid container-fluid-xl`. 

Something to maybe think about is that we can have a prop that strips all custom classes from the container, like `<ClayContainer basic/vanilla/lite />`, this will provide the defaults for 90% of the users in DXP like you said, but also give the dev that's using Clay for a personal project a single prop to enable flexibility in styling via CSS with just a single prop.

jbalsas replied:
> I didn't understand the exact goal of Clay, I thought it was a low-level library first, DXP-specific second

<img width="799" alt="Screen Shot 2020-05-11 at 09 44 02" src="https://user-images.githubusercontent.com/905006/81536536-06ea7080-936c-11ea-89f2-a0596fa87559.png">

That 👆 😂 

> `<ClayContainer basic/vanilla/lite />`

You were earlier arguing that you wouldn't bother checking props and go about it your way with CSS, so why the change of heart? 

I can understand @bryceosterhaus's struggle here, so I'm not pushing hard to make those the defaults for `<ClayContainer>`.

We need to:
- Fulfil the library's goal which is making Lexicon UIs easier
- We need to analyze what it means to build Lexicon UIs. Our biggest source of data here is DXP since that's the main Lexicon UI implementor. 
- We need to provide defaults and components based on how real developers are developing real UIs. That's the fastest way to provide real value to the users of our library.

kresimir-coko replied:
You got me there @jbalsas with that screenshot 😆 

### Why the change of heart? 
Not really a change of heart I think, if it's one prop that will disable all custom classes except `.container` I think is a good compromise. If that prop shows up on autocomplete as one of the first ones that's even better. But then again, my PoV was different when we started this discussion, as I was looking at it from a different use-case than I am now. Now I agree with you @jbalsas.

### Fluid Container
I think that having a FluidContainer on top of Container might muddy things up unnecessarily, as you said, 90% of usages are Fluid, and we can maybe add a `notFluid` prop that will omit Fluidity from the container. I would really dislike going into DXP code a couple years from now and seeing `<FluidContainer />` everywhere

wincent replied:
> @wincent, @marcoscv-work, thoughts?

This is one of those ones where I don't have anything new to add beyond what I've already said (here, [and on the issue](https://github.com/liferay/clay/issues/2421))  — and in other places about [what/why I think Clay even is](https://github.com/liferay/clay/pull/3034#issuecomment-603125574) — , and the discussion here is already very difficult to follow so I won't help much by adding more comments.

bryceosterhaus replied:
> This means that by not providing this we're refusing to provide real value to real users of our library based on some not-so-real user/library expectation. 

Yeah this totally makes sense, my thought is based more css vs react component discrepancies. **My main concern is that we should have consist naming in our components that stem from @clayui/css.** By defaulting `ClayLayout.Container` to be fluid, seemed to be at odds with @clayui/css. Also, if Lexicon is recommending the styles of `.container-fluid` to be used 90% of the time, and `.container` just 10% of the time, maybe we should just have one class and not use the non-fluid styles.

> If you want to be real purist here, let's maybe create FluidContainer that composes Container and provides the right semantics and ergonomics to our devs. 

I'm not sure I would label myself as a _purist_ 😂, I just want to make sure we are consistent from @clayui/css to our react components. I like this idea for naming though. Maybe even going with `ClayLayout.ContainerFluid` for more consistency. I think this would be inline with both @clayui/css and the need in DXP.

bryceosterhaus replied:
> @kresimir-coko 
> I would really dislike going into DXP code a couple years from now and seeing <FluidContainer /> everywhere


What would be the issue with this? I don't think it'lI be an issue really and I think it will actually be better for readability in the future and knowing exactly what the containers purpose is. I also don't mind it because it'll be easy to find/replace for any future refactors

marcoscv-work replied:
> @wincent, @marcoscv-work, thoughts?

We could create `FluidContainer` and `Container` to fit semantic but finally, we wouldn't use `Container` for anything, so I'm closer to fix the semantic in future versions of CSS than creating components or variations we don't really need

jbalsas replied:
Actually, when we weigh in our `autofit` rows and cols this might play a part since those do more or less need an inert container... so `Container` (without the "container"  class actually... 😂) could help bridge the gave between the different grids.

Or maybe something else, so we'd have

Layout (Stack)  -> Container/FluidContainer -> Row/Col
                          |-> Row autofit -> Col autofit

Adding one additional layer feels a bit over-engineered... so maybe we could simplify with:

```jsx
<Container> -> <div class="container">...</div>
<Container autofit> -> <div>...</div>
<FluidContainer> -> <div class="container-fluid container-fluid-max-xl">...</div>
```

marcoscv-work replied:
Container 'type' autofit sounds good to me, if it makes clear this is not the BS container is enough.

It is out of time but I was thinking about something more semantic which wraps everything, something like:

Using **<Grid>**
<Grid container>
<Grid row>
<Grid col>
<Grid autofit-row>
<Grid autofit-col>

wincent commented:
@jbalsas already beat me to it on the subject of cross-implementation consistency (which I think is pretty desirable), but I was going to provide a couple of links that could be useful for comparing against the taglib API:

- [`<clay:container>` attributes and class names they map to](https://github.com/john-co/liferay-portal/blob/2179cd9aba2cd1b5d726a6d3c767a0065cd880d9/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ContainerTag.java#L130-L150).
- [`<clay:col>` attributes and class names they map to](https://github.com/john-co/liferay-portal/blob/2179cd9aba2cd1b5d726a6d3c767a0065cd880d9/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ColTag.java#L166-L208).

bryceosterhaus commented:
> aligning some APIs and outputs with what we found based on real usage.

I think we should probably clarify expectations on this a bit. Are we thinking the taglibs and the components should be 1-for-1 API match? IMO I think they should be similar but not exact. If we make them exact, I think it forces clay to export odd defaults to the API that would be unexpected for a user. I made a comment about this on the container component with the container-fluid class above.

On the portal side, I think it makes sense to default values though..

```jsx
// via jsx
<Clay.Container />

// renders
<div class="container" />




// jsp
<clay:container />

// clay:container internal use
<Clay.Container fluid fluidSize="xl" />

// renders
<div class="container-fluid container-fluid-max-xl">
```


Otherwise, if we are pushing that the default Container of clay to be `container-fluid container-fluid-max-xl`, then maybe that should be the default on the css side for `.container`

Does my line of thinking make sense here or am I wildly off?

bryceosterhaus commented:
added an additional commit. Went ahead and added the `row-cols-${size}-${value}`  API to row. This will make it easy to configure a full grid. I also updated some demos and added `containerElement` prop which allows users to provide their own component or a different html element.

kresimir-coko commented on packages/clay-layout/src/Row.tsx:
```diff
@@ -0,0 +1,96 @@
+	/**
+	 * Vertical positioning of row contents
+	 */
+	align?: 'start' | 'center' | 'end';
```
`align-items` has more than these 3 options. [MDN on align-items](https://developer.mozilla.org/en-US/docs/Web/CSS/align-items)

bryceosterhaus replied:
Yeah I am following bootstraps docs on their alignment classes. https://getbootstrap.com/docs/4.0/layout/grid/#vertical-alignment

I think we can start with these 3 for now and then evaluate later if people need more.

kresimir-coko commented on packages/clay-layout/src/Row.tsx:
```diff
@@ -0,0 +1,96 @@
+	/**
+	 * Horizontal positioning of row contents
+	 */
+	justify?: 'start' | 'center' | 'end' | 'around' | 'between';
```
Same as `align-items`, has more options than the ones listed here.

bryceosterhaus replied:
Same as above comment, I am trying to keep it minimal and follow bootstraps grid docs for what should be available.

kresimir-coko commented on packages/clay-layout/stories/index.tsx:
```diff
@@ -0,0 +1,147 @@
+import ClayLayout from '../src';
+
+const exampleStyles = {
+	backgroundColor: 'rgba(86,61,124,.15)',
```
Not sure if it's too much, maybe you can provide some insight @pat270, but what do you think about using some brand colors here? Adding some branding to storybook might make it look more "ours" if you know what I mean.

pat270 replied:
This sounds good to me. Maybe #E7E7ED or #cdced9?

kresimir-coko commented on packages/demos/stories/ListPage.tsx:
```diff
@@ -296,7 +297,7 @@ export default () => {
 			</ClayManagementToolbar>
 
-			<div className="container" style={{paddingTop: 8}}>
+			<ClayLayout.Container style={{paddingTop: 8}}>
```
In the words of my toxic teacher from elementary school: "8 what? Potatoes?". Shouldn't we explicitly say pixels here?

bryceosterhaus replied:
I don't think its a huge issue for us to specify since in React we know that any style that is applied via integer becomes a px. I would agree with you in the case of a string, then we would need to specify.

https://reactjs.org/docs/dom-elements.html#style
> React will automatically append a “px” suffix to certain numeric inline style properties. If you want to use units other than “px”, specify the value as a string with the desired unit.

jbalsas commented:
Hey @bryceosterhaus!

> Are we thinking the taglibs and the components should be 1-for-1 API match? IMO I think they should be similar but not exact.

They can't really be 1-for-1 since Java-JS doesn't translate all that well. That being said, the closer the better.

> If we make them exact, I think it forces clay to export odd defaults to the API that would be unexpected for a user.

This is a different topic. I also answered in the [container-fluid](https://github.com/liferay/clay/pull/3204#discussion_r422835440) discussion. To summarize, I believe Clay should provide the defaults that make sense to achieve its main goal. Clay's goal is making it easier to build Lexicon UIs. Based on real experience building Lexicon UIs, those defaults are not in this PR.

```
// jsp
<clay:container />

// clay:container internal use
<Clay.Container fluid fluidSize="xl" />
```

One source of confusion I see here is that you seem to think `JSP` implementations actually use React components...  which will be true and not true at the same time. We currently don't have SSR, so JSPs do not defer on React components to render. They do to hydrate for interactive components, though.

> Otherwise, if we are pushing that the default Container of clay to be container-fluid container-fluid-max-xl, then maybe that should be the default on the css side for .container
>
> Does my line of thinking make sense here or am I wildly off?

I think there's a mixup with 2 concerns.
- What are the right defaults for Clay as a Library? -> Those that simplify building Lexicon UIs. Forcing a developer to always write `<Clay.Container fluid fluidSize="xl">` means the defaults we provide are not valuable since they're of no use 90% of the time. What good are defaults that only work 10% of the time?
- Defaulting `<Container>` to `.container-fluid` seems odd to you, and on that I can agree. Maybe a middle ground would be to create a highish lower level component like **FluidContainer** that composes **Container** and provides the right semantics and ergonomics to our devs.

bryceosterhaus commented:
@jbalsas I commented on the container-fluid comment as well, but I'll respond to 

> create a "highish" lower level component like FluidContainer that composes Container and provides the right semantics and ergonomics to our devs.

Yeah I think this is the best means forward for this since it keeps consistency with @clayui/css and also provides a helpful OOTB solution for dxp.


On similar note, I created an issue to further decide how to avoid these reoccurring discussions around clay + dxp. I think the confusion comes up because we have so many different layers(dxp, taglibs, react components, css) that are intertwined and don't necessarily have clear definitions to guide them. https://github.com/liferay/clay/issues/3214 **Please use this issue for continued high-level comments.**

bryceosterhaus commented:
updated with `ContainerFluid` component

jbalsas commented on packages/clay-layout/src/ContainerFluid.tsx:
```diff
@@ -0,0 +1,30 @@
+	size?: React.ComponentProps<typeof Container>['fluidSize'];
+}
+
+const ClayContainerFluid: React.FunctionComponent<IProps> = ({
```
LGTM!

nit: Should it `FluidContainer` instead of `ContainerFluid`?

bryceosterhaus replied:
Since its `.container-fluid`, it made more sense to me in keeping it consistent with that. I also prefer `NounDescriber`(if that makes sense)

jbalsas replied:
I think it's weird... because in the other components we did the opposite?

I'm thinking now at `ContentSection`, `ContentRow` and `ContentCol`...  it might just be my lack of familiarity with the grammar, but those feel like `DescriberNoun`. Basically i can tell because describer is common and noun is the thing that changes at the end.

It's true that we don't have many such cases in Clay, but others I found that also seem to contradict this are `ItemField`, `ItemText` and `ItemTitle`...

jbalsas replied:
Ah, yeah, never mind... I think I see it now 😂 👍

jbalsas replied:
Ah, yeah, never mind, I think I see it now... 😂 👍

jbalsas replied:
Or maybe not! 😂 @wincent, @pat270, some native english speakers here, please? I'm fine either way 👍

jbalsas replied:
Just to state my confusion.

`ContentSection`, `ContentRow` and `ContentTitle`... one would assume as follows that we might have a `ContentContainer`.

`ItemField`, `ItemText`, `ItemTitle`... similarly, I'd think an `ItemContainer` seems logical...

From the above, `FluidContainer` seems to follow the sequence... but I was never good at this 😂

wincent replied:
Native English speaker here... English laughs at your pathetic request to have any kind of consistent rules.

![fart-in-your-general-direction](https://user-images.githubusercontent.com/7074/81941827-9a84a100-95f9-11ea-8b36-52e9bbd26575.gif)

(I know the gif is "French", but it accurately summarizes the English attitude to consistency.)

The general grammatical pattern in English is to put modifiers (adjectives) before nouns (eg. "red car"). Things get complicated because English is quite happy to use nouns as modifers too (ie. noun like "money" being used like an adjective in the term "money bag") — but still, the modifier goes first. Unlike in Spanish, which is pretty much the opposite (eg. "coche rojo").

But programming is another layer above English, so there are other concerns beyond mere grammar. For example, we like to group related things together using a prefix. That's why `ItemField`, `ItemText`, `ItemTitle` start with a common prefix: the "thing" here is "Field", "Text", "Title", but the common relationship is that they are all the "thing" of an "Item", so we extract the common bit out and stick it at the front. And even more obviously, everything in Clay starts with "Clay", not because "Clay" is a modifier or an adjective or a noun or anything — it's the common root that it all relates to.

With respect to `ContainerFluid` vs `FluidContainer`, I have no idea. The "thing" it is is a container, and "Fluid" is the modifier that describes the container, so I'd expect it to be `FluidContainer`. But as @bryceosterhaus  said, he was being consistent with `.container-fluid`.

bryceosterhaus replied:
yeah the more I think about it I really have no idea why somethings sounds "more right" than the other way around. So I think I'll just rest my decision based on the class name itself.

jbalsas commented:
@bryceosterhaus, thanks for the update, that makes sense to me! 

At the risk of dragging this over for an additional day, do you mind weighing over at https://docs.google.com/document/d/13K-jishFtNLOoWCsWffjsCc6xsfaiH9_-ofXwoGdx3Y? (and anyone else interested, of course... looking at you, @kresimir-coko, @matuzalemsteles, @pat270...). 

I think we might be close to getting something more or less coherent that covers our most used cases that involve what we have here plus the `autofit-row` + `autofit-col` cases... plus some additional higher levels like `<Sheet>` and `<Card>` that we might or might not need to add high-level components for...

Looking for everyone to have a bit of a bigger picture to avoid too much back and forth in the implementation during the next few days, but certainly looking to wrap this up no later than this week (if possible) and continue iterating.

pat270 commented:
@bryceosterhaus The only thing that caught my eye on the storybook demos was:
```
<ClayLayout.Row>
	<ClayDemoColumn md={8} size={12}>
		{'.col-12 .col-md-8'}
	</ClayDemoColumn>
</ClayLayout.Row>
```

having the size attribute output `.col-#` seems a little odd. Bootstrap groups all the `.col-#` classes under XS, see https://getbootstrap.com/docs/4.4/layout/grid/#grid-options.

bryceosterhaus commented:
Updated with a few additional ideas. `<ClayLayout.Content*` is a component that uses "autofit-*" classes. See https://github.com/liferay/clay/pull/3204/commits/6244ab8aa35059cc681ead2c1530e6fe320eed2d#diff-58014a94aedc67ab5431dd06b8a5bf63 for what it would look like in action.

jbalsas commented:
> Updated with a few additional ideas. <ClayLayout.Content* is a component that uses "autofit-*" classes. See 6244ab8#diff-58014a94aedc67ab5431dd06b8a5bf63 for what it would look like in action.

Interesting @bryceosterhaus! I was playing around with the idea of an `autofit` attribute yesterday at https://github.com/jbalsas/liferay-portal/pull/2166.

I think this is more aligned with what @marcoscv-work envisioned... as we discussed, the problem here is that now you need to know 2 different set of APIs... but you did have to know about `autofit` anyways, so it doesn't look so different. I can be convinced and follow this to match our jsp tags...

Also, if the desire is to move away from lowish-level container-row-col to more content-aware components this seems like a good move...

@marcoscv-work, @eduardoallegrini, @wincent... thoughts?

wincent commented:
> @marcoscv-work, @eduardoallegrini, @wincent... thoughts?

I'd much rather hear from the first two people you mention, @jbalsas; I don't want to contribute any noise to an already very dense thread.

marcoscv-work commented:
> > Updated with a few additional ideas. <ClayLayout.Content* is a component that uses "autofit-*" classes. See [6244ab8](https://github.com/liferay/clay/commit/6244ab8aa35059cc681ead2c1530e6fe320eed2d)#diff-58014a94aedc67ab5431dd06b8a5bf63 for what it would look like in action.
> 
> Interesting @bryceosterhaus! I was playing around with the idea of an `autofit` attribute yesterday at [jbalsas/liferay-portal#2166](https://github.com/jbalsas/liferay-portal/pull/2166).
> 
> I think this is more aligned with what @marcoscv-work envisioned... as we discussed, the problem here is that now you need to know 2 different set of APIs... but you did have to know about `autofit` anyways, so it doesn't look so different. I can be convinced and follow this to match our jsp tags...
> 
> Also, if the desire is to move away from lowish-level container-row-col to more content-aware components this seems like a good move...
> 
> @marcoscv-work, @eduardoallegrini, @wincent... thoughts?

That's closer to what I tried to explain, it looks more semantic and I really like it.
If we agree we can use the pattern in clay and Liferay portal :-D

jbalsas commented:
> That's closer to what I tried to explain, it looks more semantic and I really like it.
If we agree we can use the pattern in clay and Liferay portal :-D

I'll start moving in that direction, then

jbalsas commented on packages/clay-layout/src/Content.tsx:
```diff
@@ -0,0 +1,129 @@
+		<ContainerElement
+			{...otherProps}
+			className={classNames(className, 'autofit-col', {
+				'autofit-col-expand': expand,
```
Looking at usages in DXP, `autofit-col-end` is also possible and very extended. It seems to be we usually have either nothing, `expand` or `end`. The `gutters` value is compatible with the above. 

We might want to add `end` as another prop... or maybe there's a better more semantic name for this...

@pat270, @marcoscv-work?  What's the difference between nothing, `expand` and `end`? How would we need to name an enumeration of this?

pat270 replied:
@jbalsas `expand` is a good name.

We shouldn't include `autofit-col-end`. It was deprecated at https://github.com/liferay/clay/issues/803. The replacement for it is using `autofit-float-end` on `autofit-row`.

---

```
<div class="autofit-float-end autofit-row">
  <div class="autofit-col autofit-col-expand">
    <div class="autofit-section">
      The quick brown fox jumped over the lazy dog.
    </div>
  </div>
  <div class="autofit-col">
    <div class="autofit-section">
      <button>Add</button>
    </div>
  </div>
</div>
```

If the `autofit-col`s don't fit on one line, the last item will break to new line and everything (`autofit-col`) will be aligned right.

---

```
<div class="autofit-float-end-{sm|md}-down autofit-row"></div>
```
At the breakpoint `{sm|md}`, if `autofit-col`s don't fit on one line, last item will break to new line and aligned right.

---

```
<div class="autofit-float autofit-row"></div>
```

If doesn't fit on one line, breaks to new line and aligned left.

---

```
<div class="autofit-float-{sm|md}-down autofit-row"></div>
```

At breakpoint `{sm|md}`, if doesn't fit on one line, breaks to new line and aligned left.

---

- `autofit-col` sizes the column to be as wide as its content.

- `autofit-col autofit-col-expand` forces the column to fill the remaining space. If there are multiple `autofit-col-expand`s then the column widths are supposed to be divided equally. This doesn't work in all situations right now. I think we can fix this.

We also have:

```
<ul class="card-page card-page-equal-height">
  <li class="card-page-item col-{breakpoint}-#"></li>
  <li class="card-page-item col-{breakpoint}-#"></li>
  <li class="card-page-item col-{breakpoint}-#"></li>
</ul>
```

I don't think we should include this use case here, but just getting it out. We also have a grid for cards `card-page` and `card-page card-page-equal-height`. You can see a demo at https://pat270.github.io/clay3-test-site/vclay-table-dd/content/test-card-type-template/.

jbalsas replied:
> It was deprecated at #803

That was like 2 years ago! 😂 We will need to discuss also how to deprecate things in claycss in the future :D 

@bryceosterhaus, that means if I understand correctly that `expand` on col is okay, but we're missing `float` or something like that on [ContentRow](https://github.com/liferay/clay/pull/3204/files#diff-3d9a224ed0f982f449f890b93b0b359aR47)?

wincent replied:
> You can see a demo at https://pat270.github.io/clay3-test-site/vclay-table-dd/content/test-card-type-template/.

(Off topic, but) there are some funny flickering on that page in Chrome 81:

![Screenshot 2020-05-18 -115816-2fAZXhEY](https://user-images.githubusercontent.com/7074/82200524-67932380-98ff-11ea-935d-7b515d9765ab.gif)

Takes a while to trigger it, but you see it when I move the mouse off of the rightmost card.

pat270 replied:
Thanks @wincent, not sure when this started happening (I'm blaming a Chrome update on this). Any `transition-timing-function` other than `ease-in` or `ease-in-out` fixes this artifact.

jbalsas commented on packages/clay-layout/src/Sheet.tsx:
```diff
@@ -0,0 +1,106 @@
+			className={classNames(
+				className,
+				'sheet-footer',
+				'sheet-footer-btn-block-sm-down'
```
Is this always needed? This actually feels oddly specific. I can't find any usage of this in DXP... but @pat270 can call me out on this 😉

pat270 replied:
It was a specific request for https://projects.invisionapp.com/d/main#/projects/prototypes/12432242. It hasn't been deprecated. If we don't need it, I'm good with removing it if Lexicon approves.
<img width="111" alt="sheet-block-btn" src="https://user-images.githubusercontent.com/788266/82079719-277e3780-9698-11ea-8faa-f025b46cdd40.png">

jbalsas replied:
I meant if this class was really a default one, and always ncessary.

We might need to keep it, I’m just not sure if it’s a default... I’ll take a closer look at DXP next week, but we might just pass it through ‘className’ for now if in doubt...

pat270 replied:
OH I see. I would say don't default to `sheet-footer sheet-footer-btn-block-sm-down`. It's very specific to that form.

bryceosterhaus commented:
Updated to add float

jbalsas commented on packages/clay-layout/src/index.tsx:
```diff
@@ -0,0 +1,33 @@
+	ContentRow,
+	ContentSection,
+	Row,
+	Sheet,
```
What's the reason behind grouping all `Sheet` components but not all `Content` ones?

bryceosterhaus replied:
Thats my bad, I will update it. The thought behind it is that we don't want multiple namespaces, so instead of `<ClayLayout.Sheet.Header>` it should be `<ClayLayout.SheetHeader`. So generally, the only namespace should be the top level component, in this case `ClayLayout`

bryceosterhaus replied:
Just updated

jbalsas commented on packages/clay-layout/src/Content.tsx:
```diff
@@ -0,0 +1,147 @@
+				[`autofit-float-${float}`]: typeof float === 'string',
+				'autofit-padded': padded,
+				'autofit-padded-no-gutters': noGutters === true,
+				[`autofit-padded-no-gutters-${noGutters}`]:
```
Hey @pat270, one doubt here... is it `autofit-padded` always needed with `autofit-padded-no-gutters`? Is there any logic here that needs to be coded?

Meaning, is just `autofit-padded-no-gutters` correct? or setting `-no-gutters` implies `padded`?

@bryceosterhaus, just in case those 2 props are in fact related...

pat270 replied:
@jbalsas @bryceosterhaus 
No `autofit-padded` isn't always needed with `autofit-padded-no-gutters`. I added the same padding to both because it gets confusing. You can use `autofit-padded autofit-padded-no-gutters` or just `autofit-padded-no-gutters`.

I think the most common use case would be `autofit-row autofit-padded-no-gutters-x` which uses negative margins to align the left and right side of `autofit-row` (similar to Bootstrap's `row`), but provides a little white space top and bottom. 

---

`autofit-padded` adds `padding: 0.25rem 0.5rem` on `autofit-col`

---

`autofit-padded-no-gutters-x` adds `padding: 0.25rem 0.5rem` on `autofit-col` and `margin-left: -0.5rem; margin-right: -0.5rem` on `autofit-row`

---

`autofit-padded-no-gutters` adds `padding: 0.25rem 0.5rem` on `autofit-col` and `margin: -0.25rem -0.5rem` on `autofit-row`

jbalsas commented on packages/clay-layout/src/Content.tsx:
```diff
@@ -0,0 +1,147 @@
+	 * Provides the benefit of aligning content via flexbox without losing the behavior
+	 * of floated elements at the expense of extra markup.
+	 */
+	float?:
```
Not sure what I'll name this in Java since `float` is a reserved keyword 😂

jbalsas commented:
Hey @bryceosterhaus, this looks almost good to go to me...

I just added a question about the [`padded` and `noGutters` relation](https://github.com/liferay/clay/pull/3204#discussion_r426158603) and a couple more comments here and there about APIs we might not want to commit to without more data.

My suggestion would be to remove those props for now, and go through [LPS-111257 Analyze Layout Strategies used in React Components in DXP - Report](https://docs.google.com/spreadsheets/d/1040_UfBNNj2wzoBjhsoec872Ns2Uvf4MiFbQuUoMRqg) based on an initial release (or local) of this. Keep in mind that I think that's missing `autofit` and `sheet` usage, so those should be added too.

With that, we should have first-hand visibility on these APIs and their ergonomics. We can create some LPSs to send updated Apps to their teams using the new components.

What do you think?

bryceosterhaus commented:
> My suggestion would be to remove those props for now, 

Are we sure we want to remove them? `*-padded` and `*-padded-no-*` are used at least 20 times in DXP from my count. Also, it seems from the conversation that there isn't really any coding needed to link the two classes together since they behave independently.

bryceosterhaus commented:
> We can create some LPSs to send updated Apps to their teams using the new components.

Yeah I am up for this. I think it'd be wise to get an initial release of `@clayui/layout` merged into portal first and then we can go ahead and make these updates and LPSs.

jbalsas commented:
> Are we sure we want to remove them? *-padded and *-padded-no-* are used at least 20 times in DXP from my count. Also, it seems from the conversation that there isn't really any coding needed to link the two classes together since they behave independently.

No, I meant the `container`, `formSize` and `order`. The `padded` ones make sense to me and are pretty straightforward.

jbalsas commented:
> Yeah I am up for this. I think it'd be wise to get an initial release of @clayui/layout merged into portal first and then we can go ahead and make these updates and LPSs.

SGTM

bryceosterhaus commented:
> No, I meant the view, formSize and order. The padded ones make sense to me and are pretty straightforward.

Ah gotcha, I just did a quick search and it looks like view would be used 28 times and formSIze would also be used 28. `order` looks to be used much less.

Does it make sense to keep the props for those 28 cases of each? I would lean towards yes, unless you are thinking they will be removed.

jbalsas commented:
> Does it make sense to keep the props for those 28 cases of each? I would lean towards yes, unless you are thinking they will be removed.

Yeah, for 28 cases it definitely makes sense... I was just trying to err on the side of caution and add them later once we saw how a minimal API looked like. Just trying to avoid a "breaking change" if at some point we realized one of those was unused or weird. If it's clear to you, go for it, I don't have a particular concern about this.

bryceosterhaus commented:
> I was just trying to err on the side of caution and add them later once we saw how a minimal API looked like.

Totally makes sense. I think these APIs won't be too problematic for us and given the number of uses, I think it makes sense to add.

I went ahead and updated the PR:
- removed `order` prop since it wasn't being used much
- removed the sample uses of @clayui/layout in other components. Once the package is released I will go through and update our existing packages.
- squashed into a single commit

Not sure If I missed anything else here in this PR, I think I covered everything though. I will mark this as ready to review.

jbalsas commented on packages/clay-card/src/CardWithNavigation.tsx:
```diff
@@ -74,41 +74,56 @@ export const ClayCardWithNavigation: React.FunctionComponent<IProps> = ({
-				)}
-			</ClayCard.Body>
+							{title && (
+								<span className="autofit-col autofit-col-expand">
```
This would be a good place to use `ContentCol` 😂

bryceosterhaus replied:
Yeah I will go through and update our other components once a cut is released. I think if we update it before releasing @clayui/layout then CI might break for other PRs

jbalsas replied:
Sounds good! 👍

jbalsas commented on packages/clay-card/src/CardWithNavigation.tsx:
```diff
@@ -74,41 +74,56 @@ export const ClayCardWithNavigation: React.FunctionComponent<IProps> = ({
-						</span>
-						<span className="autofit-col autofit-col-expand">
-							<span className="autofit-section">
+			{(horizontal || title || description) && (
```
I'm looking at this in GitHub, so this might just be @wincent's most beloved UI playing tricks, but how did this change make it into this PR? Was it always here?

jbalsas replied:
Yeah... probably just appears in the diff because we merged upstream... :sigh:

wincent replied:
![E5R-Rwjp_400x400](https://user-images.githubusercontent.com/7074/82432156-ecad4280-9a8f-11ea-9097-3e5cea371f11.jpg)

jbalsas commented on packages/clay-layout/package.json:
```diff
@@ -0,0 +1,39 @@
+{
+	"name": "@clayui/layout",
+	"version": "3.0.0-alpha.1",
+	"description": "ClayLayout component",
```
nit, for the future: go over the descriptions in our packages. Doesn't feel like this says much about the package, what it does or why you'd want to download. Maybe we can get some copy from the Lexicon site and/or team

wincent replied:
One of the consequences of dividing a library into 45 packages is that you have to come up with 45 descriptions. 😀

jbalsas commented on packages/clay-layout/src/Col.tsx:
```diff
@@ -0,0 +1,88 @@
+	/**
+	 * Position in which to align column
+	 */
+	align?: 'start' | 'center' | 'end';
```
I'm still unsure about `align`... do we also have usages for this? I'm fine to keep it if you think it's safe and useful.

jbalsas commented on packages/clay-layout/src/Row.tsx:
```diff
@@ -0,0 +1,96 @@
+}
+
+const ClayRow: React.FunctionComponent<IProps> = ({
+	align,
```
Same thoughts I posted about `Col`

jbalsas commented on packages/clay-layout/src/Row.tsx:
```diff
@@ -0,0 +1,96 @@
+	 * Number of columns the row should contain before breaking
+	 * on to a new line
+	 */
+	numOfCol?:
```
Do we want to commit to this API?

bryceosterhaus commented:
@jbalsas I'll verify some of these props in portal just to make sure. And yeah, I plan on adding the docs once this PR gets merged and before release tomorrow.
